docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1975 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #1.    def _update_is_cands(Q, T, μ_Q, σ_Q, M_T, Σ_T, min_dist, indices, is_cands):
</pre><p>I think how/where you've chosen to refactor feels rather strange and even the name of the function is so vague that I think we should go back to what we had before. We've now refactored to a point where the collective code is losing meaning. What do you think?</p><p> </p><p>Note that reverting is not a bad thing! This means that we are converging on something that feels "right" and additional changes are making it "worst". It is good for you to "see" this (i.e., we should never feel pressured to go too far). </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Reply:
> the name of the function is so vague 

Not only that, but also the variables of the function seem disconnected as well (just my opinion: for instance, `indices`, or even `is_cands`!!)  However, I thought this is a private function and I know what I am doing. Apparently my thought-process was not flawless! I can see that it doesn't clearly deliver the message of its purpose and I just thought that is okay as it private. Will change back its structure.

Reply:
@seanlaw 
I know you told me to avoid refactoring as you didn't like the function and the arguments. So, I thought why not refactoring the first two functions `_select_candidates` and `_prune_candidates(...)` (see below)

```
def _prune_candidates_general(T, m, M_T, Σ_T, min_dist, is_cands, direction='right'):
    n = T.shape[0] 
    k = n - m + 1 # number of subsequences
    
    T_subseq_isfinite = np.isfinite(M_T)
    
    
    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
    
    for i in np.flatnonzero(T_subseq_isfinite):
        if np.all(is_cands == False):
            break
        cands_idx = np.flatnonzero(is_cands) #to update `cands_idx` 
        
        # refactoring!
        if direction == 'left':
            non_trivial_cands_idx = cands_idx[cands_idx < max(0, i - excl_zone)]
        elif direction == 'right':
            non_trivial_cands_idx = cands_idx[cands_idx > i + excl_zone]
        
        if len(non_trivial_cands_idx) == 0:
            continue 
        
        
        cand_idx_chunks = _get_chunks_ranges(non_trivial_cands_idx) # get left_non_trivial cand indices in chunks
        for (start, stop) in cand_idx_chunks:
            QT = core.sliding_dot_product(T[i:i+m], T[start: (stop+m-1)]) #stop + m - 1 is correct for converting... 
            #stop from "subsequence index" space to "time series index" space... 
            #Do not replace with the incorrect stop - m + 1 in this case!
            D = core._mass(T[i:i+m], T[start: (stop+m-1)], QT, M_T[i], Σ_T[i], M_T[start:stop], Σ_T[start:stop])

            mask = np.flatnonzero(D < min_dist)   
            is_cands[start:stop][mask] = False

            if len(mask):
                is_cands[i] = False
        
        
    return is_cands
```
I tested it on all cases studies and two more `np.random` case.

Reply:
@seanlaw 
Also, I noticed we should have updated `cands_idx` inside the `for-loop` of `_prune_candidates(...)`; because we are updating `is_cands` in the `for-loop` BUT we were using the same `cands_idx`. (I think that was the reason the running time was sometimes very high )

Reply:
Cool. This works! Please replace the last two function and push your commits - I have some suggestions but it'll be easier for me to comment inline. Maybe just call it `_find_candidates`?